File: challenge_utils.js
import fs from "fs";

export function readInput(challengePath) {
  return fs.readFileSync(`${challengePath}/input.txt`, "utf-8");
}

export function readInputLines(challengePath) {
  return readInput(challengePath).split("\n");
}

export function readInputInts(challengePath) {
  return readInputLines(challengePath).map((line) => parseInt(line));
}

export function slidingWindows(array, length) {
  const windows = [];
  for (let i = 0; i < array.length - (length - 1); i++) {
    windows.push(array.slice(i, i + length));
  }
  return windows;
}

export function sum(array) {
  return array.reduce((acc, x) => acc + x, 0);
}

export function countIf(array, predicate) {
  let count = 0;
  let previousElem;
  array.forEach((elem) => {
    if (predicate(elem, previousElem)) {
      count += 1;
    }
    previousElem = elem;
  });
  return count;
}

File: challenge_utils.py
import os


def read_input(file):
    challenge_dir = os.path.dirname(os.path.abspath(file))
    with open(os.path.join(challenge_dir, "input.txt")) as f:
        return f.read()


def read_input_lines(file):
    return read_input(file).split("\n")


def read_input_ints(file):
    return [int(line) for line in read_input_lines(file)]


def sliding_windows(array, length):
    windows = []
    for i in range(len(array) - (length - 1)):
        windows.append(array[i : i + length])
    return windows


def count_if(array, predicate):
    count = 0
    previous_elem = None
    for elem in array:
        if predicate(elem, previous_elem):
            count += 1
        previous_elem = elem
    return count
